File: packages/cli/src/commands/embed.ts
import type { Command } from 'commander';
import chalk from 'chalk';
import ora from 'ora';
import { isConfigured, runEmbedPipeline, hasRepos, type EmbedProgress } from '@repog/core';

/**
 * Options passed from command line.
 */
interface EmbedOptions {
  includeFileTree?: boolean;
  verbose?: boolean;
  batchSize?: number;
}

/**
 * Run the embed command in default mode (with spinner).
 */
async function runDefaultMode(options: EmbedOptions): Promise<void> {
  const spinner = ora('Generating embeddings...').start();

  let lastProgress: EmbedProgress | null = null;

  // Validate and warn about batch size
  let batchSize = options.batchSize;
  if (batchSize !== undefined && batchSize > 100) {
    console.log(chalk.yellow(`Warning: Batch size ${batchSize} exceeds maximum, capping at 100`));
    batchSize = 100;
  }

  try {
    for await (const progress of runEmbedPipeline({
      includeFileTree: options.includeFileTree ?? false,
      batchSize,
    })) {
      lastProgress = progress;

      if (progress.type === 'batch') {
        const total = progress.totalChunks ?? '?';
        spinner.text = `Generating embeddings... ${progress.chunksEmbedded}/${total} chunks`;
      }
      // repo_skip and error events are silently tracked in default mode
    }

    spinner.succeed(chalk.green('Embeddings generated'));

    // Print summary
    if (lastProgress) {
      printSummary(lastProgress);
    }
  } catch (error) {
    spinner.fail(chalk.red('Embedding failed'));
    if (error instanceof Error) {
      console.error(chalk.red(`  Error: ${error.message}`));
    }
    process.exit(1);
  }
}

/**
 * Run the embed command in verbose mode (no spinner).
 */
async function runVerboseMode(options: EmbedOptions): Promise<void> {
  console.log(chalk.bold('\nGenerating embeddings...\n'));

  let lastProgress: EmbedProgress | null = null;

  // Validate and warn about batch size
  let batchSize = options.batchSize;
  if (batchSize !== undefined && batchSize > 100) {
    console.log(chalk.yellow(`Warning: Batch size ${batchSize} exceeds maximum, capping at 100`));
    batchSize = 100;
  }

  try {
    for await (const progress of runEmbedPipeline({
      includeFileTree: options.includeFileTree ?? false,
      batchSize,
    })) {
      lastProgress = progress;

      switch (progress.type) {
        case 'batch':
          console.log(
            chalk.green(
              `  + Batch ${progress.batchIndex}/${progress.batchTotal} - ${progress.chunksEmbedded} chunks embedded`
            )
          );

          // Report individual chunk errors in this batch if any
          if (progress.batchErrors && progress.batchErrors.length > 0) {
            progress.batchErrors.forEach((err) => {
              console.log(chalk.red(`    x Chunk ${err.chunkId}: ${err.error}`));
            });
          }
          break;

        case 'repo_skip':
          console.log(chalk.gray(`  ~ ${progress.repoFullName} (already embedded)`));
          break;

        case 'error':
          console.log(chalk.red(`  x Error: ${progress.errorMessage}`));
          break;
      }
    }

    console.log('');

    // Print summary
    if (lastProgress) {
      printSummary(lastProgress);
    }
  } catch (error) {
    console.error(chalk.red('\nEmbedding failed'));
    if (error instanceof Error) {
      console.error(chalk.red(`  Error: ${error.message}`));
    }
    process.exit(1);
  }
}

/**
 * Print the final summary.
 */
function printSummary(progress: EmbedProgress): void {
  console.log(chalk.bold('\nSummary:'));
  console.log(`  Chunks embedded: ${chalk.green(progress.chunksEmbedded ?? 0)}`);

  if ((progress.chunksSkipped ?? 0) > 0) {
    console.log(`  Chunks skipped:  ${chalk.gray(progress.chunksSkipped)} (repos unchanged)`);
  }

  if ((progress.chunksErrored ?? 0) > 0) {
    console.log(`  Chunks errored:  ${chalk.red(progress.chunksErrored)}`);
  }

  console.log('');
}

/**
 * Run the embed command logic.
 */
async function runEmbed(options: EmbedOptions): Promise<void> {
  // Check if configured
  if (!isConfigured()) {
    console.log(chalk.red('Run `repog init` first.'));
    process.exit(1);
  }

  // Check if any repos exist
  if (!hasRepos()) {
    console.log(chalk.red('No repositories found. Run `repog sync` first.'));
    process.exit(1);
  }

  if (options.verbose) {
    await runVerboseMode(options);
  } else {
    await runDefaultMode(options);
  }
}

/**
 * Register the `repog embed` command.
 * Generates embeddings for repository content.
 */
export function register(program: Command): void {
  program
    .command('embed')
    .description('Generate embeddings for repository content')
    .option('--include-file-tree', 'Include file tree chunks in embeddings', false)
    .option('-v, --verbose', 'Show detailed progress', false)
    .option('--batch-size <number>', 'Batch size for embedding requests (default: 20, max: 100)', (value) => {
      const parsed = parseInt(value, 10);
      if (isNaN(parsed) || parsed < 1) {
        throw new Error('Batch size must be a positive integer');
      }
      return parsed;
    })
    .action(async (options: EmbedOptions) => {
      await runEmbed(options);
    });
}

File: commands/embed.go
package commands

import (
	"context"
	"fmt"
	"os"
	"time"

	"github.com/briandowns/spinner"
	"github.com/fatih/color"
	"github.com/spf13/cobra"

	"github.com/hackastak/repog/internal/config"
	"github.com/hackastak/repog/internal/db"
	"github.com/hackastak/repog/internal/embed"
)

var embedCmd = &cobra.Command{
	Use:   "embed",
	Short: "Generate embeddings for synced repositories",
	Long:  "Generate vector embeddings for repository chunks using Google Gemini.",
	RunE:  runEmbed,
}

var (
	embedIncludeFileTree bool
	embedVerbose         bool
	embedBatchSize       int
)

func init() {
	embedCmd.Flags().BoolVar(&embedIncludeFileTree, "include-file-tree", true, "Include file_tree chunks in embeddings (default true)")
	embedCmd.Flags().BoolVar(&embedVerbose, "verbose", false, "Show detailed progress")
	embedCmd.Flags().IntVar(&embedBatchSize, "batch-size", 20, "Embedding batch size (max 100)")
}

func runEmbed(cmd *cobra.Command, args []string) error {
	green := color.New(color.FgGreen).SprintFunc()
	yellow := color.New(color.FgYellow).SprintFunc()
	red := color.New(color.FgRed).SprintFunc()

	// Load config
	cfg, err := config.LoadConfig()
	if err != nil {
		fmt.Fprintln(os.Stderr, red("Run `repog init` first."))
		os.Exit(1)
	}

	geminiKey, err := config.GetGeminiAPIKey()
	if err != nil {
		fmt.Fprintln(os.Stderr, red("Run `repog init` first."))
		os.Exit(1)
	}

	// Open database
	database, err := db.Open(cfg.DBPath)
	if err != nil {
		fmt.Fprintln(os.Stderr, red("Database error:"), err)
		os.Exit(1)
	}
	defer func() { _ = database.Close() }()

	// Start embedding pipeline
	var s *spinner.Spinner
	if !embedVerbose {
		s = spinner.New(spinner.CharSets[14], 100*time.Millisecond)
		s.Suffix = " Generating embeddings..."
		s.Start()
	}

	eventCh := embed.RunEmbedPipeline(context.Background(), embed.EmbedOptions{
		IncludeFileTree: embedIncludeFileTree,
		BatchSize:       embedBatchSize,
		DB:              database,
		GeminiAPIKey:    geminiKey,
	})

	for event := range eventCh {
		switch event.Type {
		case "batch":
			if embedVerbose {
				fmt.Printf("Batch %d/%d: %d embedded, %d skipped, %d errors\n",
					event.BatchIndex, event.BatchTotal, event.ChunksEmbedded, event.ChunksSkipped, event.ChunksErrored)
				// Show first unique error from this batch
				if len(event.Errors) > 0 {
					fmt.Println(red("  Error:"), event.Errors[0])
				}
			} else if s != nil {
				progress := float64(event.BatchIndex) / float64(event.BatchTotal) * 100
				s.Suffix = fmt.Sprintf(" Generating embeddings... %.0f%% (%d/%d chunks)",
					progress, event.ChunksEmbedded, event.TotalChunks)
			}

		case "repo_skip":
			if embedVerbose {
				fmt.Println(yellow("~"), "skipped", event.RepoFullName)
			}

		case "error":
			if embedVerbose {
				fmt.Println(red("✗"), "error:", event.RepoFullName)
			}

		case "done":
			if s != nil {
				s.Stop()
			}

			if event.ChunksErrored > 0 {
				fmt.Println(yellow("✓"), fmt.Sprintf("Embedding complete — %d embedded, %d skipped, %d errors",
					event.ChunksEmbedded, event.ChunksSkipped, event.ChunksErrored))
			} else {
				fmt.Println(green("✓"), fmt.Sprintf("Embedding complete — %d embedded, %d skipped, %d errors",
					event.ChunksEmbedded, event.ChunksSkipped, event.ChunksErrored))
			}
		}
	}

	return nil
}
